### PR TITLE
WIP: rework IntegerRange math impl

### PR DESCRIPTION
**Pull request by staabm**
refs https://github.com/phpstan/phpstan-src/pull/652

leaving the work in progress in this PR for now, so I can continue work from a different computer

### CODE REVIEW

**Comment by staabm** [comment]
@ondrejmirtes my local testing shows me that math.php tests are now passing. some of the initially-added integer-range-math tests in `integer-range-types.php` are failling though.

are you comfortable enough with the newly added failling tests, that I could/should drop those in `integer-range-types.php` - I feel these are mostly redundant with your new tests...

**Comment by ondrejmirtes** [comment]
@staabm Please don't delete any tests :) If you want to adjust the asserts if you feel like the new ones make more sense then fine, but don't delete any asserts...

**Comment by staabm** [comment]
@ondrejmirtes thx integer-range related tests I created with the first iteration of integer-math and the one you added pass now.

could you help me interpreting the remaining errors.. I am not sure whether these are related to my PR or how to fix them.

File: src/File/ParentDirectoryRelativePathHelper.php [hunk 1]
```diff
@@ -54,6 +54,11 @@ public function getFilenameParts(string $filename): array
 		}
 
 		$dotsCount = $parentPartsCount - $i;
+
+		if ($dotsCount < 0) {
```
Comment (staabm) [inline src/File/ParentDirectoryRelativePathHelper.php:58]: I had to re-add these lines, which were deleted with https://github.com/phpstan/phpstan-src/pull/637/files#r696555714

it seems it was a false-positive that phpstan could detect this case with the new integer-range math

File: tests/PHPStan/Analyser/data/integer-range-types.php [hunk 2]
```diff
@@ -228,10 +228,10 @@ public function math($i, $j, $z, $pi, $r1, $r2, $rMin, $rMax, $x, $y) {
...
 		assertType('int<5, max>', $j * $rMax);
 
-		assertType('int<-19, 13>', $r1 + $z);
-		assertType('int<-2, 30>', $r1 - $z);
-		assertType('int<-20, 30>', $r1 * $z);
-		assertType('float', $r1 / $z);
+		assertType('int<-19, -10>|int<2, 13>', $r1 + $z);
+		assertType('int<-2, 9>|int<21, 30>', $r1 - $z);
+		assertType('int<-200, -20>|int<1, 30>', $r1 * $z);
+		assertType('float|int<0, 10>', $r1 / $z);
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/integer-range-types.php:234]: after I added coverage for the test-cases in https://github.com/phpstan/phpstan-src/pull/652 I realized that these operations here can be expressed more precily. and therefore we have the changes here

File: tests/PHPStan/Analyser/data/integer-range-types.php [hunk 3]
```diff
@@ -240,7 +240,7 @@ public function math($i, $j, $z, $pi, $r1, $r2, $rMin, $rMax, $x, $y) {
 		assertType('int<2, max>', $pi * 2);
 		assertType('float|int<0, max>', $pi / 2);
 		assertType('int<2, max>', 1 + $pi);
-		assertType('int<1, max>', 2 - $pi);
+		assertType('int<min, 2>', 2 - $pi);
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/integer-range-types.php:243]: a bug in the previous expectations

File: src/Analyser/MutatingScope.php [hunk 4]
```diff
@@ -1267,96 +1267,143 @@ private function resolveType(Expr $node): Type
...
-						} elseif ($type instanceof ConstantIntegerType) {
-							if ($node instanceof Node\Expr\BinaryOp\Minus || $node instanceof Node\Expr\AssignOp\Minus ||
-								$node instanceof Node\Expr\BinaryOp\Div || $node instanceof Node\Expr\AssignOp\Div) {
-								$leftMin = max($leftMin, $type->getValue());
-								$leftMax = $leftMax !== null ? min($leftMax, $type->getValue()) : $type->getValue();
+				/**
+				 * @param PhpParser\Node\Expr\AssignOp|PhpParser\Node\Expr\BinaryOp\BinaryOp $node
+				 * @param IntegerRangeType|ConstantIntegerType|UnionType $operand
+				 */
+				$math = static function (IntegerRangeType $range, $node, $operand) use (&$math): Type {
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1274]: Callback referencing itself like this leads to memory leak explosion. Please have a private method instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
I'm gonna look into the remaining failures.

**Comment by ondrejmirtes** [comment]
Alright, this should be ready. Let's see what CI says 🤞

**Comment by ondrejmirtes** [comment]
@staabm Please check the commits to see what I had to do.

**Comment by ondrejmirtes** [comment]
I'm happy with this, thanks for the help!

**Comment by staabm** [comment]
> I'm happy with this, thanks for the help!

thank you for getting it in shape to land it!

**Comment by staabm** [comment]
>  Please check the commits to see what I had to do.

the commits itself logic logical to me. I was not able make sense of the test-errors and derive the necessary changes from it.

hopefully with more experience of the codebase this will get better over time.

**Pull request merged**